### ISSUE TITLE
.NET 8 preview7 Blazor WASM AAD authentication ends with There was an error trying to log you in: '"undefined" is not valid JSON'

### ISSUE THREAD

**Issue by markushaslinger** [dotnet/aspnetcore#49956]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

I am getting the same issue as described in #44981 again: after coming back from login at Azure AD the user sees 'There was an error trying to log you in: '"undefined" is not valid JSON'' at the 'authentication/login-failed' route.

However, after manually navigating around the token is available so it kind-of worked anyway.

Same as with the previous issue, adding the following solves the issue, so I suspect a similar cause:

```
<ItemGroup>
    <TrimmerRootAssembly Include="Microsoft.AspNetCore.Components.WebAssembly.Authentication" />
</ItemGroup>
```

Also: error does not appear during development, only after publish also pointing to the linker

### Expected Behavior

No error should be displayed after the redirect if the operation was successful.

### Steps To Reproduce

Standard Blazor WASM application using `Microsoft.Authentication.WebAssembly.Msal` (8.0.0-preview.7.23375.9), set up to use 'redirect' and not the 'pop up' auth option.

### .NET Version

8.0.100-preview.7.23376.3

**Comment by hectorm-bmg** [comment]
I'm getting a similar behavior ("undefined" is not valid JSON') but in this case I'm using Azure ADB2C with an IdentityServer4 external provider. 
I tried using 
<ItemGroup>
    <TrimmerRootAssembly Include="Microsoft.Authentication.WebAssembly.Msal" />
</ItemGroup>

but the issue is still there. Any suggestions would be appreciated.

**Comment by mkArtakMSFT** [comment]
@halter73 isn't this something you've addressed already (your PR to the some AAD library I think) ?

**Comment by hakenr** [comment]
In the `8.0.0` release:
* This issue appears to be a regression from #44981 (cc @javiercn).
* It results in a "LogInFailed" UI without any visible notifications.
* No errors are logged in the browser console or in the `RemoteAuthenticatorView.LogInFailed` context (notably, downgrading MSAL to `7.0.13` yields an error: `"undefined" is not valid JSON`).
* Disabling trimming for `Microsoft.Authentication.WebAssembly.Msal` when publishing the app resolves the issue immediately.

I believe this should not be classified as "technical debt" for future ".NET 9 Planning", but rather recognized as a regression that could potentially cause significant issues for users.

**Comment by kyleherzog** [comment]
I experienced the same issue when upgrading to .NET 8.0 with Azure B2C.  I had to add the following to get it to work.
```
  <ItemGroup>
    <TrimmerRootAssembly Include="Microsoft.Authentication.WebAssembly.Msal" />
    <TrimmerRootAssembly Include="Microsoft.AspNetCore.Components.WebAssembly.Authentication" />
  </ItemGroup>
```
Specifying just Microsoft.Authentication.WebAssembly.Msal alone did not work for me.

**Comment by hakenr** [comment]
@kyleherzog Sorry for the confusion. Actually I do have both the MSAL and WebAssembly.Authentication rooted and it is probably the `Microsoft.AspNetCore.Components.WebAssembly.Authentication` which is the important one for this symptom.

**Comment by czlatea** [comment]
I can confirm. The issue is solved in our case by mentioning only one assembly
`	<ItemGroup>
		<TrimmerRootAssembly Include="Microsoft.AspNetCore.Components.WebAssembly.Authentication" />
	</ItemGroup>`

**Comment by YingXiaoMing** [comment]
i have same problem,but i don't know how to fix it. nomatter i add 
<TrimmerRootAssembly Include="Microsoft.AspNetCore.Components.WebAssembly.Authentication" />
but it doesn't work

**Comment by jumulr** [comment]
Using Net8 1.0, TrimmerRootAssembly does not fix the issue.
Please help.

**Comment by czlatea** [comment]
@martinblampied could you please check if you need both assemblies listed ? For me, it worked nicely listing only **Microsoft.AspNetCore.Components.WebAssembly.Authentication**

**Comment by martinblampied** [comment]
@czlatea Doesn't work for me; tried both and just Authentication in both Client and Server projects using .net 8

**Comment by SeanLeitzinger** [comment]
Same issue on .NET 8 and I have tried adding both to the trimmer exception. Pushing this to .NET 9 is unacceptable. This has been going on long enough. It's been years now.

**Comment by Olsenss** [comment]
Had the same issue. Going from .NET 7 -> .NET 8, I had to add the following in the Client csproj:
<ItemGroup>
	<TrimmerRootAssembly Include="Microsoft.Authentication.WebAssembly.Msal" />
	**<TrimmerRootAssembly Include="Microsoft.AspNetCore.Components.WebAssembly.Authentication" />**
</ItemGroup>

This resolved the issue.

Another thing I noticed was that it was only in Microsoft Edge the problem occured, in Google Chrome there was never an issue.

**Comment by MaddMugsy** [comment]
In .NET 8 and have the same issue.  The TrimmerRootAssembly fix did not work for me.  The problem occurs in both Chrome and Edge.

Worth noting though that it doesn't happen locally, only in release mode in a deployed environment.

**Comment by meronz** [comment]
I have tried with both adding TrimmerRootAssembly and disabling trimming altogether with PublishTrimmed = false and the problem persists.

This is showstopping bug happening in a critical piece of an LTS release. Postponing the fix to .NET 9 seems absurd to me, given that these workarounds don't work for everybody.

**Comment by acmesoft-arg** [comment]
In my case, with blazor wasm (not hosted) didn't work even configure:
```
<ItemGroup>
	<TrimmerRootAssembly Include="Microsoft.Authentication.WebAssembly.Msal" />
	<TrimmerRootAssembly Include="Microsoft.AspNetCore.Components.WebAssembly.Authentication" />
</ItemGroup>
```
in the project file.
I commented this line to make it work in release (but the app lost the redirect functionality): 
`//options.ProviderOptions.LoginMode = "redirect";`

**Comment by hakenr** [comment]
To all who report that the trimming setting does not address their issue – it's probable you're encountering a different problem. A key symptom of this issue is: it only appears after **publishing** your project (which is when assembly trimming occurs). If you're facing authentication problems in a development environment (like when running from Visual Studio), you're dealing with a separate issue, and the `TrimmerRootAssembly` setting will not assist you.

**Comment by SimonMouridsen** [comment]
> To all who report that the trimming setting does not address their issue – it's probable you're encountering a different problem. A key symptom of this issue is: it only appears after **publishing** your project (which is when assembly trimming occurs). If you're facing authentication problems in a development environment (like when running from Visual Studio), you're dealing with a separate issue, and the `TrimmerRootAssembly` setting will not assist you.

The fix below does not help me, and I only experience the issue when deploying (Azure App Service). Had no issue with .NET 7.
 
    <ItemGroup>
            <TrimmerRootAssembly Include="Microsoft.Authentication.WebAssembly.Msal" />
	    <TrimmerRootAssembly Include="Microsoft.AspNetCore.Components.WebAssembly.Authentication" />
    </ItemGroup>

**Comment by KptNorad** [comment]
yes with <PublishTrimmed>false</PublishTrimmed> the problem is gone. Even it was a hard way to test it. I had to clean all obj folders manually. Before that mono was out of sync and a cleanup did not work.

**Comment by SimonMouridsen** [comment]
> yes with `<PublishTrimmed>false</PublishTrimmed>` the problem is gone. Even it was a hard way to test it. I had to clean all obj folders manually. Before that mono was out of sync and a cleanup did not work.
> 
> so I tried
> 
> ```
> <ItemGroup>
>         <TrimmerRootAssembly Include="Microsoft.Authentication.WebAssembly.Msal" />
>     <TrimmerRootAssembly clude="Microsoft.AspNetCore.Components.WebAssembly.Authentication"/>
> </ItemGroup>
> ```
> 
> with trimming again and delete all obj folders. And it is working.

I did not try `<PublishTrimmed>false</PublishTrimmed>`, but I did make sure this time to clean all relevant bin+obj folders - now deploying works for me with only one assembly "rooted":

```
  <ItemGroup>
    <TrimmerRootAssembly Include="Microsoft.AspNetCore.Components.WebAssembly.Authentication"/>
  </ItemGroup>
```

**Comment by syazdian** [comment]
> Had the same issue. Going from .NET 7 -> .NET 8, I had to add the following in the Client csproj:
> 
> ```
> <ItemGroup>
> 	<TrimmerRootAssembly Include="Microsoft.Authentication.WebAssembly.Msal" />
> 	<TrimmerRootAssembly Include="Microsoft.AspNetCore.Components.WebAssembly.Authentication" />
> </ItemGroup>
> ```
> 
> This resolved the issue.
> 
> Another thing I noticed was that it was only in Microsoft Edge the problem occured, in Google Chrome there was never an issue.

Thanks! you were right A) it was only in Edge B) it fixed my issue

**Comment by MaddMugsy** [comment]
Quick update on my experience:  We have the Msal nuget package as a transitive dependency, buried fairly deep in some class libraries.  Adding the TrimmerRootAssembly for WebAssembly.Authentication and/or PublishTrimmed = false to that library wasn't sufficient.  

It did work however once we added TrimmerRootAssembly to that library and every library that referenced it, along with the actual Blazor WASM project.

This might be a little overkill, but it'll do the trick for now.

**Comment by Rajesh-Goel** [comment]
<PublishTrimmed>false</PublishTrimmed> this worked. 

    With below settings

    <ItemGroup>
            <TrimmerRootAssembly Include="Microsoft.Authentication.WebAssembly.Msal" />
        <TrimmerRootAssembly Include="Microsoft.AspNetCore.Components.WebAssembly.Authentication"/>
    </ItemGroup>

**Comment by luisabreu** [comment]
When changing the csproj file (with one of the previous entries), don't forget to delete the bin and obj folders...

**Comment by rbvpowell** [comment]
I was still getting this issue after trying the TrimmerRootAssembly fix and the PublishTrimmed false fixes, but my build pipeline was running an incremental build so the previously trimmed assemblies were still getting used.
Try disabling incremental build or cleaning your build environments if you are also getting this issue despite trying the other fixes.

I've opted to disable trimming entirely as I was additionally getting an issue with Telerik components that `<PublishTrimmed>false</PublishTrimmed>` also fixed. I don't know how many other issues this could be causing in my dependencies so I don't trust having assembly trimming enabled.

(I also tested reverting the changes and it did bring back both of my issues)

**Comment by mwasson74** [comment]
Are there any updates on this? I am also facing this issue with my .NET 8 application deployed to an Azure AppService using Azure ADB2C.

I have a hosted Blazor WASM application. In which project(s) do I put the `<PublishTrimmed>false</PublishTrimmed>` to see if that works for me?

**Comment by KptNorad** [comment]
> Are there any updates on this? I am also facing this issue with my .NET 8 application deployed to an Azure AppService using Azure ADB2C.
> 
> I have a hosted Blazor WASM application. In which project(s) do I put the `<PublishTrimmed>false</PublishTrimmed>` to see if that works for me?

you have to put it in the wasm project and please don't forget to delete the bin and obj folders for all projects before you deploy

**Comment by stian-bay** [comment]
Just setup a new project on the hosted wasm model and I had to put this in the client project for the authentication to work for azure AD. On the Microsoft Identity Platform it worked without this.

```
  <ItemGroup>
    <TrimmerRootAssembly Include="Microsoft.AspNetCore.Components.WebAssembly.Authentication" />
  </ItemGroup>
```

**Comment by mwasson74** [comment]
> please don't forget to delete the bin and obj folders for all projects before you deploy

This doesn’t really apply in a CI/CD setup does it? I run `dotnet publish` in my build, publish the artifacts from the `publish` folder, and then my release deploys those artifacts to the Azure AppService.

**Comment by antoprd** [comment]
I am facing this issue on a IIS hosted Blazor WASM application.

**Comment by davispg** [comment]
Thankful I found this thread. Experienced the issue upgrading our .NET 7.0 Blazor WASM PWA to .NET 8.0. Applied the suggested changes, which alongside removing the bin/obj folders and a full rebuild worked a treat.
`<ItemGroup>
<TrimmerRootAssembly Include="Microsoft.Authentication.WebAssembly.Msal" />
<TrimmerRootAssembly Include="Microsoft.AspNetCore.Components.WebAssembly.Authentication"/>
</ItemGroup>`

**Comment by aherrick** [comment]
why is this still an issue in .NET 8? yes there is this fix but come on

**Comment by halter73** [comment]
I'm having trouble reproducing the `There was an error trying to log you in: '"undefined" is not valid JSON` even after publishing https://github.com/halter73/BlazorHostedWasmAAD with the following commands:

```
> git clean -xdf
Removing .vs/
Removing Client/bin/
Removing Client/obj/
Removing Server/bin/
Removing Server/obj/
Removing Shared/bin/
Removing Shared/obj/
> cd Server
> dotnet publish -c Release
MSBuild version 17.8.3+195e7f5a3 for .NET
  Determining projects to restore...
  Restored H:\temp\WasmAAD\Shared\WasmAAD.Shared.csproj (in 132 ms).
  Restored H:\temp\WasmAAD\Client\WasmAAD.Client.csproj (in 228 ms).
  Restored H:\temp\WasmAAD\Server\WasmAAD.Server.csproj (in 229 ms).
  WasmAAD.Shared -> H:\temp\WasmAAD\Shared\bin\Release\net8.0\WasmAAD.Shared.dll
  WasmAAD.Client -> H:\temp\WasmAAD\Client\bin\Release\net8.0\WasmAAD.Client.dll
  WasmAAD.Client (Blazor output) -> H:\temp\WasmAAD\Client\bin\Release\net8.0\wwwroot
  WasmAAD.Server -> H:\temp\WasmAAD\Server\bin\Release\net8.0\WasmAAD.Server.dll
  Optimizing assemblies for size may change the behavior of the app. Be sure to test after publishing. See: https:
  //aka.ms/dotnet-illink
  Optimizing assemblies for size. This process might take a while.
  WasmAAD.Server -> H:\temp\WasmAAD\Server\bin\Release\net8.0\publish\
> cd .\bin\Release\net8.0\publish\
> dotnet .\WasmAAD.Server.dll --urls https://localhost:7189
info: Microsoft.Hosting.Lifetime[14]
      Now listening on: https://localhost:7189
info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Production
info: Microsoft.Hosting.Lifetime[0]
      Content root path: H:\temp\WasmAAD\Server\bin\Release\net8.0\publish
info: Microsoft.Hosting.Lifetime[0]
      Application is shutting down... 
```

I then navigate to https://localhost:7189 with a private browsing window and log in successfully without any errors in either the server logs or client developer console.

![Home page while logged in](https://github.com/dotnet/aspnetcore/assets/54385/2a563fc4-e594-4c7e-8420-e94ab218940f)

To create the BlazorHostedWasmAAD project, I did the following:

- Run `dotnet new blazorwasm -au SingleOrg -ho` using the .NET 7.0.406 SDK as specified in global.json
- Followed the guide at https://learn.microsoft.com/en-us/aspnet/core/blazor/security/webassembly/hosted-with-microsoft-entra-id?view=aspnetcore-7.0#use-of-an-azure-active-directory-b2c-tenant to configure it with an Azure AD B2C tenant
- Then updated the TFMs from net7.0 to net8.0 and the NuGet packages to their latest 8.0 versions.

You can see each step in the projects [commits](https://github.com/halter73/BlazorHostedWasmAAD/commits/main/). I did not have to use `TrimmerRootAssembly` or `PublishTrimmed` in my csproj.

I see that there are a lot of people have run into this issue, so I'm sure I'm missing something. Can anyone figure out what I may doing wrong?

**Comment by markushaslinger** [comment]
@halter73 I forked your repo and set it up like my project, relevant changes:

- Redirect Mode is set: https://github.com/markushaslinger/BlazorHostedWasmAAD_repro/commit/786e08e64e1b6e1ba0c23144cdffad520bf5a1a4 
- It's not hosted by a .NET process but I just throw it up on a Nginx which hosts it
- Instead of `options.ProviderOptions.DefaultAccessTokenScopes.Add("https://{TENANT}.onmicrosoft.com/{SERVER API APP CLIENT ID}/API.Access");` I use `options.ProviderOptions.DefaultAccessTokenScopes.Add("{SERVER API APP CLIENT ID}/.default");` (I honestly can't remember why any more)

After these changes (+ entering my own tenant and client of course) and running `dotnet publish -c Release` (afaik the `Release` tag is now the default anyway, but I'm so used to writing it) I threw the `wwwroot` content on the server, tried the login, and:

![image](https://github.com/dotnet/aspnetcore/assets/1212708/dbe832c3-6a0b-4870-a2ff-0d30b02bdfd2)

Don't think my setup is totally unusal, but please tell me if I'm doing something I'm not supposed to do (the redirect is a requirement though, can't change that). I hope that helps you to reproduce the issue!

**Comment by mattfrear** [comment]
FWIW (to help reproduce) - I had this issue after following the steps here https://learn.microsoft.com/en-us/aspnet/core/blazor/security/webassembly/standalone-with-microsoft-entra-id

Solution runs fine on my localhost but shows the JS error once deployed to Azure.

The same workaround works for me.
```xml
<ItemGroup>
    <TrimmerRootAssembly Include="Microsoft.AspNetCore.Components.WebAssembly.Authentication"/>
</ItemGroup>
```

**Comment by mattfrear** [comment]
> I see that there are a lot of people have run into this issue, so I'm sure I'm missing something. Can anyone figure out what I may doing wrong?

Hi @halter73. Many of us in this thread only experienced the issue when we deployed to Azure, localhost worked fine.

FWIW (to help reproduce) - I had this issue after following the steps here https://learn.microsoft.com/en-us/aspnet/core/blazor/security/webassembly/standalone-with-microsoft-entra-id

Solution runs fine on my localhost but shows the JS error once deployed to Azure.

The same workaround works for me.
```xml
<ItemGroup>
    <TrimmerRootAssembly Include="Microsoft.AspNetCore.Components.WebAssembly.Authentication"/>
</ItemGroup>
```

**Comment by milkyjoe90** [comment]
Is this expected to be folded in to the 8.0.4 release?

**Comment by Eddie-Hartman** [comment]
For anyone following along, I had a similar issue here #54515 that I was hoping would be fixed with 8.0.4. I updated and that does not seem to be the case.

Can we stop closing tickets until something is confirmed as fixed AND released? I even asked in my other issue how I'd confirm that this works locally so I don't have to do a full build and deploy, but no dice.

**Comment by mcschwa** [comment]
pls fix

**Comment by DylanLyon** [comment]
Hoping this gets fixed soon. Loading app state from an encoded URL (like in #54514 ) has been broken for months in our app and is a pretty basic/critical feature to be broken for so long.

**Comment by ldabrowskiairspan** [comment]
Key functionality (authentication) of a framework (in this case Blazor) being shipped completely broken out of the box in an LTS release when you've known about it since preview builds is completely unacceptable, closing an issue when it is NOT fixed, and expecting users to jump through hoops to find a CLOSED GitHub issue with a workaround is even worse. Doing things like that, alongside other shortcomings of auth in Blazor (like using a deprecated version of [oidc-client-js](https://github.com/IdentityModel/oidc-client-js)) basically ensures Blazor will never see serious adoption.

**Comment by milkyjoe90** [comment]
It looks like the fix might be included in the upcoming 8.0.5, which based on the release cadence so far, I'd guestimate that it'll be out on May 7th

https://github.com/dotnet/aspnetcore/pull/54655

You're right though, they really need to do something about the underlying oidc-client, there hasn't been a release of it since 2021 and the repo was archived in March 2022, the replacement [oidc-client-ts](https://github.com/authts/oidc-client-ts) is actively maintained, regularly updated and well supported - they really do need to do something.

**Comment by Eddie-Hartman** [comment]
@milkyjoe90 afaik patches come out the second Tuesday of each month (but are not guaranteed), so that would be May 14th. Takes a bit for things to propagate (like packages getting posted to NuGet and what not), so I'd just check the 15th.

**Comment by milkyjoe90** [comment]
> @milkyjoe90 afaik patches come out the second Tuesday of each month (but are not guaranteed), so that would be May 14th. Takes a bit for things to propagate (like packages getting posted to NuGet and what not), so I'd just check the 15th.

Ahhh Patch Tuesday, of course! Can’t believe I missed that! Lines up perfectly with all the other MS patches that get released!

**Comment by michaelhannes** [comment]
Can anyone confirm if I am having the same issue here ? 

Essentially everything works well overall, especially running locally, however when we deploy to azure the following occurs: 

1) login page redirects to B2C login when user clicks "Login" button. 

2) B2C login page, everything as expected

3) on "return" to our login page, we check   "if (B2Cuser.Identity.IsAuthenticated)"  and then Navigate to "/user) 

when running locally it runs fast and no refresh issues or anything. 

when deployed to Azure - after B2C login redirect, we are return to index.razor ("/") and the user is authenticated but does not Navigate to "/user" 

however, if we manually refresh the page it successfully checks (B2Cuser.Identity.IsAuthenticated) and Navigates to "/user" 

I am in process of adding the TrimmerRootAssembly(s) for Msal and Authentication

**Comment by stian-bay** [comment]
Did not work for me on 8.0.4 but it works now after updating to 8.0.5 packages 🎉

Aslo tested if the returnUrl is fixed but this does not redirect to the return Url (RedirectToLogin.razor)
```
#@inject NavigationManager Navigation

@code {
    protected override void OnInitialized()
    {
        Navigation.NavigateTo($"authentication/login?returnUrl={Uri.EscapeDataString(Navigation.Uri)}");
    }
}

```
So I still have to have this workaround in RedirectToLogin.razor for preserving the returnUrl and redirecting on login
```
@using Microsoft.Extensions.Options
@using Microsoft.AspNetCore.Components.WebAssembly.Authentication

@inject NavigationManager Navigation

@inject IOptionsSnapshot<RemoteAuthenticationOptions<ApiAuthorizationProviderOptions>> Options

@code {
    protected override void OnInitialized()
    {
        Navigation.NavigateToLogin(Options.Get(Microsoft.Extensions.Options.Options.DefaultName).AuthenticationPaths.LogInPath);
    }
}
```

**Comment by curved-gulden** [comment]
> To all who report that the trimming setting does not address their issue – it's probable you're encountering a different problem. A key symptom of this issue is: it only appears after **publishing** your project (which is when assembly trimming occurs). If you're facing authentication problems in a development environment (like when running from Visual Studio), you're dealing with a separate issue, and the `TrimmerRootAssembly` setting will not assist you.

That is not completely true. My issue only appears after publishing, yet adding the TrimmerRootAssembly still does not work. None of the solutions in any of the comments or issues do. This is a complete show-stopper. Login is pretty critical.

**Comment by mattfrear** [comment]
Cool, fixed for me using Microsoft.Authentication.WebAssembly.Msal 8.0.7, I've removed the TrimmerRootAssembly workaround.

**Comment by imsam67** [comment]
I created a new Blazor WASM app that uses Azure AD B2C using `dotnet new blazorwasm -au IndividualB2C --aad-b2c-instance "{AAD B2C INSTANCE}" --client-id "{CLIENT ID}" --domain "{TENANT DOMAIN}" -o {PROJECT NAME} -ssp "{SIGN UP OR SIGN IN POLICY}"` command as suggested in this article: https://learn.microsoft.com/en-us/aspnet/core/blazor/security/webassembly/standalone-with-azure-active-directory-b2c?view=aspnetcore-8.0

Everything works fine locally. I'm hosting the app on Azure using Static Web App service and clicking the Login link does nothing, except for throwing this error: 

![Screenshot 2024-09-15 151832](https://github.com/user-attachments/assets/34aaf870-17a3-4b1f-a639-93cd087294f5)

If basic login doesn't work, this is a show stopper for us to use Blazor WASM!

**Comment by mattfrear** [comment]
@imsam67 have you upgraded Microsoft.Authentication.WebAssembly.Msal NuGet package to the latest version?

**Comment by imsam67** [comment]
I'm using version 8.0.8, same as the WebAssembly package version.
